### ISSUE TITLE
@WithSecurityContext test without MockMVC

### ISSUE THREAD

**Issue by poznachowski** [spring-projects/spring-security#3721]
I have a Spring Boot application (REST-based using JAX-RS) and couple of endpoints secured with a custom `AuthenticationProvider`. For integration testing I'm not using MockMVC, but REST-assured.
How can I make use of `@WithSecurityContext` annotation to mock my `SecurityContext`?

**Comment by rwinch** [comment]
@poznachowski I don't think this is currently possible. Does REST-assured have some hook like MockMvc's `RequestPostProcessor`?

**Comment by jgoldhammer** [comment]
Yes, it has...
https://github.com/jayway/rest-assured/wiki/Usage#bootstrapping-restassuredmockmvc

**Comment by rwinch** [comment]
@jgoldhammer Thanks for the response. This appears to be rest assured with Spring MVC. The request is to support JAX-RS. Is there something I'm missing?

**Comment by jgoldhammer** [comment]
Yeah, you are right...  I could not understand why anyone will use Jaxrs instead of Spring MVC with Spring Boot... :-)

**Comment by poznachowski** [comment]
Sorry for lack of response. As stated REST-assured has only Spring MVC support included. Afaik, there is no way to make it work in a similar fashion, as REST-assured will be solely used as HTTP client.

About choosing JAX-RS over Spring MVC.. I welcomed Spring ecosystem quite recently, coming from a custom-made stack of Guice, JAX-RS, Jackson etc. (way before Dropwizard became popular). Hence, everything in Spring was new and I was tempted to stick with JAX-RS (+ Swagger works flawlessly). To be honest, I still like the syntax more. The choice can be considered a little short-sighted, as it introduced different set of issues (like the one described here), but it helped understand Spring codebase better.

Back to the topic. I created custom test configuration along with custom annotation to allow mocking out security in integration tests.

**Comment by rwinch** [comment]
There is nothing wrong with using JAX-RS (after all, Spring is about choice).

I'd be glad to accept a pull request for support on this, but I'm not sure that I will have time for this feature in the near future.

**Comment by spring-issuemaster** [comment]
If you would like us to look at this issue, please provide the requested information. If the information is not provided within the next 7 days this issue will be closed.

**Comment by xxxxxxjun** [comment]
I'd like to pick this up.

Before writing anything I want to make sure I'm aiming at the right target. #18812 was closed because documenting the separate-thread case wasn't the use case raised here, so I'd rather confirm the scope first.

Two questions:

1. By "JAX-RS tests can run in-memory", do you mean something like Jersey's `InMemoryTestContainerFactory`, where the request never reaches the servlet layer? I want to reproduce the setup you have in mind.

2. If the goal is to make `@WithSecurityContext` usable outside MockMvc, is a framework-neutral hook in scope? The bridging currently lives in `SecurityMockMvcRequestPostProcessors` as a private `TestSecurityContextRepository`, so nothing outside MockMvc can reuse it. Exposing an equivalent in `spring-security-test` would let any servlet-based test harness opt in, without Spring Security taking on a JAX-RS dependency.

Happy to go with a docs-only change if that's closer to what you had in mind.
